src/charm.py

```diff
@@ -50,6 +56,13 @@ def __init__(self, *args):
 
         self._init_postgresql_relation()
 
+    def _on_mongodb_client_relation_changed(self, event: ops.framework.EventBase) -> None:
+        """Handle the on MongoDB relation changed event."""
+        if "mongodb" not in self._stored.reldata:
+            self._stored.reldata["mongodb"] = {}
+        self._stored.reldata["mongodb"].update(self.mongodb.fetch_relation_data()[event.relation.id])
+        self._configure_workload(event)
```

Comment:
Just advice: typically, database credentials are specified in the configuration file.
It means `replan_services` function will not restart your application when the `endpoints_changed` emited.
I would add the following code before ` self._configure_workload(event)` line.
```suggestion
        container = self.unit.get_container(CONTAINER_NAME)
        services = container.get_services("gunicorn")
        if services and services["gunicorn"].is_running():
            container.stop("gunicorn")
        self._configure_workload(event)
```

Reply:
I've take another approach, only running _configure_workload if state has changed. Thanks for the hint.